Edge WebView2: add possibility to create app modules for apps hosting WebView2 interface

josephsl opened the pull request:
### Link to issue number:
Closes #16705 

### Summary of the issue:
Apps hosting Edge WebView2 show up as websites. However, in apps such as new Outlook (olk.exe), it is desirable to disable browse mode by default or offer custom commands and workarounds.

### Description of user facing changes
NVDA will recognize and work with app modules for apps hosting Edge WebView2 controls such as new Outlook.

### Description of development approach
Similar to wwahost and javaw, a new host app module (msedgewebview2) is created to fetch actual names of aps hosting the webview controls. The module borrows heavily from javaw app module except it is limited to fetching parent process information. Instead of importing appModuleHandler.AppModule, module-level __getattr__ is defined to return the default app module.

### Testing strategy:
Manual testing:

1. Download and install an app that embeds WebView2 controls such as new Outlook (olk) and LinkedIn, or if using Windows 11, launch Widgets (Windows+W).
2. Without the PR applied, generate develoepr info (NVDA+F1) and observe the appModule line (it will say "msedgewebview2" for app name).
3. With the PR applied, generate developer info, and observe that the parent process is now the app name ("olk" for new Outlook, for example).
4. Create an app module for one of the host apps such as olk.py for new Outlook.
5. Restart NVDA, switch to the app for te just-created app module, then generate developer info (app name and app module name will say the actual app name).

### Known issues with pull request:
None

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

<!-- Please keep the following -->
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added support for applications utilizing the Edge WebView2 runtime interface, enhancing app module creation capabilities.
  
- **Documentation**
  - Updated developer guide with additional details and examples for creating app modules for apps using Edge WebView2. Clarifications on naming conventions and subclassing were also included.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Thanks.

josephsl committed:
apModues/msedgewebview2: support loading custom app modules for apps hosting WebView2 runtime to display ther interfaces. Re #16705.

Similar approach to javaw.py: look up the parent process for WebView2 app and return the parent process name for use by NVDA when loading app modules for apps hosting WebView2 runtime.

josephsl committed:
Edge WebView2: add type hints for 'getAppNameFromHost' function. Re #16705

josephsl committed:
Edge WebView2: return the default app module if no custom app module is loaded. Re #16705.

Using module-level __getattr__, load the default app module class if no custom app module is found.

josephsl committed:
Changelog: add msedgewebview2 entry under 'changes for develoeprs' section. Re #16705

josephsl commented on user_docs/en/changes.md:
```diff
@@ -79,6 +79,7 @@ It is especially useful to read the error location markers in tracebacks. (#1632
   * You can use `scheduleThread.scheduleDailyJobAtStartUp` to automatically schedule a job that happens after NVDA starts, and every 24 hours after that.
   Jobs are scheduled with a delay to avoid conflicts.
   * `scheduleThread.scheduleDailyJob` and `scheduleJob` can be used to schedule jobs at custom times, where a `JobClashError` will be raised on a known job scheduling clash.
+* It is now possible to create app modules for apps hosting Edge WebView2 (msedgewebview2.exe) controls. (#16705, @josephsl)
```
Almost - perhaps use only the first sentence as per other changes.md entries (remember to put the issue number and GitHub username mentions for third-party contributors).

josephsl commented:
@CodeRabbitAI, read my initial comment again as the walkthrough is not equal to what I actually wrote. For reference, it does oposite to what you think the PR is doing.

josephsl commented:
@CodeRabbitAI summarize the pull request based on my latest comment about the workflow

LeonarddeR commented:
I think I'm missing how this appModule ought to be used. May be add an example first, e.g. for Outlook?

josephsl commented:
Hi, for Outlook and other apps embedding WebView2, one would create app modules just like other apps as is done for apps powered by say, wwahost and javaw. Thanks.

josephsl commented:
Hi,

One more thing I should have done: add notes about msedgewebview2.exe in dev guide - I'll take care of this soon.

Thanks.

josephsl committed:
Developer guide: mention msedgewebview2.exe and apps hosting it. Re #16705.

Note from Leonard de Ruijter: give an example of an app hosting WebView2 runtime. For this one, a third app module example is defined to showcase WebView2 and disabling browse mode by default (example: olk.exe for new Outlook).

seanbudd commented on projectDocs/dev/developerGuide/developerGuide.md:
```diff
@@ -472,12 +472,13 @@ As with other examples in this guide, remember to delete the created app module
+These include javaw.exe for running various Java programs, wwahost.exe for some apps in Windows 8 and later, and msedgewebview2.exe for displaying web-like interface on apps employing Edge WebView2 runtime.
 
-If an app runs inside a host executable, the name of the app module must be the name as defined by the host executable, which can be found through AppModule.appName property.
+If an app runs inside a host executable or employs a different app to display the interface, the name of the app module must be the name as defined by the host or the interface executable, which can be found through AppModule.appName property.
```
```suggestion
If an app runs inside a host executable or employs a different app to display the interface, the name of the app module must be the name as defined by the host or the interface executable, which can be found through the `AppModule.appName` property.
```

seanbudd commented on projectDocs/dev/developerGuide/developerGuide.md:
```diff
@@ -504,6 +505,24 @@ As a built-in app module, wwahost can be imported from nvdaBuiltin.appModules.
+    
+    --- end ---
+
+Browse mode is disabled for this example because apps employing WebView2 display their interfaces as webpages. You can remove "disableBrowseModeByDefault" line if you would like to let users navigate the app using browse mode commands.
```
```suggestion
Browse mode is disabled for this example because apps employing WebView2 display their interfaces as webpages.
You can remove the "disableBrowseModeByDefault" line if you would like to let users navigate the app using browse mode commands.
```

seanbudd committed:
Apply suggestions from code review

seanbudd commented:
@coderabbitai review
